src/verifai/samplers/multi_armed_bandit.py

```diff
@@ -0,0 +1,203 @@
+        self.cont_sampler, self.disc_sampler = None, None
+        self.rand_sampler = None
+        for subsampler in self.split_sampler.samplers:
+            if isinstance(subsampler, ContinuousMultiArmedBanditSampler):
+                assert self.cont_sampler is None
+                self.cont_sampler = subsampler
+            elif isinstance(subsampler, DiscreteMultiArmedBanditSampler):
+                assert self.disc_sampler is None
+                self.disc_sampler = subsampler
+            else:
+                assert isinstance(subsampler, RandomSampler)
+                assert self.rand_sampler is None
+                self.rand_sampler = subsampler
+
+    def nextSample(self, feedback=None):
+        return self.split_sampler.nextSample(feedback)
+
+    def getSample(self):
+        return self.split_sampler.getSample()
+
+    def update(self, sample, info, rho):
+        self.split_sampler.update(sample, info, rho)
+
+class ContinuousMultiArmedBanditSampler(BoxSampler, MultiObjectiveSampler):
+    def __init__(self, domain, alpha, thres,
+                 buckets=10, dist=None, restart_every=100):
+        super().__init__(domain)
+        if isinstance(buckets, int):
+            buckets = np.ones(self.dimension) * buckets
+        elif len(buckets) > 1:
+            assert len(buckets) == self.dimension
+        else:
+            buckets = np.ones(self.dimension) * buckets[0]
+        if dist is not None:
+            assert (len(dist) == len(buckets))
+        if dist is None:
+            dist = np.array([np.ones(int(b))/b for b in buckets])
+        self.buckets = buckets
+        self.dist = dist
+        self.alpha = alpha
+        self.thres = thres
+        self.current_sample = None
+        self.counts = np.array([np.ones(int(b)) for b in buckets])
+        self.errors = np.array([np.zeros(int(b)) for b in buckets])
+        self.t = 1
+        self.counterexamples = dict()
+        self.is_multi = False
+        self.invalid = np.array([np.zeros(int(b)) for b in buckets])
+        self.monitor = None
+        self.rho_values = []
+        self.restart_every = restart_every
+
+    def nextVector(self, feedback=None):
+        self.update(None, self.current_sample, feedback)
+        return self.generateSample()
+    
+    def generateSample(self):
+        proportions = self.errors / self.counts
+        Q = proportions + np.sqrt(2 / self.counts * np.log(self.t))
+        # choose the bucket with the highest "goodness" value, breaking ties randomly.
+        bucket_samples = np.array([np.random.choice(np.flatnonzero(np.isclose(Q[i], Q[i].max())))
+            for i in range(len(self.buckets))])
+        self.current_sample = bucket_samples
+        ret = tuple(np.random.uniform(bs, bs+1.)/b for b, bs
+              in zip(self.buckets, bucket_samples))
+        return ret, bucket_samples
+    
+    def update(self, sample, info, rho):
+        if rho is None:
```

Comment:
This should not be possible. `update` should only be called after a sample has been evaluated.